azurelinuxagent/common/logcollector.py

```diff
@@ -0,0 +1,340 @@
+    "/var/log/waagent*",
+]
+
+_FILE_SIZE_LIMIT = 30 * 1024 * 1024  # 30 MB
+_UNCOMPRESSED_ARCHIVE_SIZE_LIMIT = 150 * 1024 * 1024  # 150 MB
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class LogCollector(object):
+
+    _TRUNCATED_FILE_PREFIX = "truncated_"
+
+    def __init__(self, manifest_file_path):
+        self._manifest_file_path = manifest_file_path
+        self._must_collect_files = self._expand_must_collect_files()
+        self._create_base_dirs()
+        self._set_logger()
+
+    @staticmethod
+    def _create_base_dirs():
+        mkdir(_TRUNCATED_FILES_DIR)
+        mkdir(_LOG_COLLECTOR_DIR)
+
+    @staticmethod
+    def _set_logger():
+        _f_handler = logging.FileHandler(_OUTPUT_RESULTS_FILE_PATH, encoding="utf-8")
+        _f_format = logging.Formatter(fmt='%(asctime)s %(levelname)s %(message)s',
+                                      datefmt='%Y-%m-%d %H:%M:%S')
+        _f_handler.setFormatter(_f_format)
+        _LOGGER.addHandler(_f_handler)
+        _LOGGER.setLevel(logging.INFO)
+
+    @staticmethod
+    def _run_shell_command(command, stdout=subprocess.PIPE, output=False):
+        def format_command(cmd):
+            return " ".join(cmd) if isinstance(cmd, list) else command
+
+        def _encode_command_output(output):
+            return ustr(output, encoding="utf-8", errors="backslashreplace")
+
+        try:
+            process = subprocess.Popen(command, stdin=subprocess.PIPE, stdout=stdout, stderr=subprocess.PIPE, shell=False)
+            stdout, stderr = process.communicate()
+            return_code = process.returncode
+        except Exception as e:
+            error_msg = u"Command [{0}] raised unexpected exception: [{1}]".format(format_command(command), ustr(e))
+            _LOGGER.error(error_msg)
+            return
+
+        if return_code != 0:
+            encoded_stdout = _encode_command_output(stdout)
+            encoded_stderr = _encode_command_output(stderr)
+            error_msg = "Command: [{0}], return code: [{1}], stdout: [{2}] stderr: [{3}]".format(format_command(command),
+                                                                                                 return_code,
+                                                                                                 encoded_stdout,
+                                                                                                 encoded_stderr)
+            _LOGGER.error(error_msg)
+            return
+
+        if output:
+            msg = "Output of command [{0}]:\n{1}".format(format_command(command), _encode_command_output(stdout))
+            _LOGGER.info(msg)
+
+    @staticmethod
+    def _expand_must_collect_files():
+        # Match the regexes from the MUST_COLLECT_FILES list to existing file paths on disk.
+        manifest = []
+        for path in _MUST_COLLECT_FILES:
+            manifest.extend(sorted(glob.glob(path)))
+
+        return manifest
+
+    @staticmethod
+    def _parametrize_filepath(path):
```

Comment:
your code would be simpler and less prone to mistakes if you allow something like %lib_dir%/foo.xml instead of /var/lib/foo.xml. then when you read the manifest you can expand %lib_dir% instead of having to remember that one needs to do this parameterization each time a file path is used

Reply:
Done, thanks.